src/messaging/ExchangeHolder.h

```diff
@@ -0,0 +1,136 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/Optional.h>
+#include <lib/support/IntrusiveList.h>
+#include <messaging/ExchangeContext.h>
+
+namespace chip {
+namespace Messaging {
+
+/** @brief
+ *      This provides a wrapper around ExchangeContext that automatically manages
+ *      cleaning up the EC without any extra involvement. This is meant to be used
+ *      by application and protocol logic code that would otherwise need to closely
+ *      manange their internal pointers to an ExchangeContext and correctly
+ *      null-it out/abort it depending on the circumstances. This relies on clear rules
+ *      established by ExchangeContext and the transfer of ownership at various points
+ *      in its lifetime.
+ *
+ *      It does this by listening in on OnExchangeClosing and looking at the various
+ *      states the exchange might be in to decide amongst a couple of different tactics.
```

Comment:
Please describe the heuristics. Please replace "tactics" with "heuristics"

Reply:
The heuristics are provided  in `AbortIfNeeded`. Will reference that section here.